File: backend/node.py
from functools import wraps
from flask import Flask, jsonify, request
from wallet import Wallet
from flask_cors import CORS
from blockchain import Blockchain
app = Flask(__name__)
CORS(app)


def require_json(f):
    @wraps(f)
    def decorated_function(*args, **kwargs):
        if not request.is_json:
            response = {
                "error": "Unsupported Media Type",
                "message": "Content-Type must be application/json"
            }
            return jsonify(response), 415
        return f(*args, **kwargs)
    return decorated_function

@app.route('/wallet', methods=['POST'])
def create_keys():
    wallet.create_keys()
    if wallet.save_keys():
        global blockchain
        blockchain = Blockchain(wallet.public_key,port)
        response = {
            'public_key': wallet.public_key,
            'private_key': wallet.private_key,
            'funds': blockchain.get_balance()
        }
        return jsonify(response), 201

    response = {
        'message': "Saving the keys failed."
    }
    return jsonify(response), 500


@app.route('/wallet', methods=['GET'])
def load_keys():
    if wallet.load_keys():
        global blockchain
        blockchain = Blockchain(wallet.public_key,port)
        response = {
            'public_key': wallet.public_key,
            'private_key': wallet.private_key,
            'funds': blockchain.get_balance()
        }
        return jsonify(response), 200

    response = {
        'message': "Saving the keys failed."
    }
    return jsonify(response), 500


@app.route('/balance', methods=['GET'])
def get_balance():
    balance = blockchain.get_balance()
    if balance != None:
        response = {
            'message': 'Fetched balance Successfully.',
            'funds': balance
        }
        return jsonify(response), 200

    response = {
        'message': 'Loading wallet is faulled.',
        'wallet_set_up': wallet.public_key != None
    }
    return jsonify(response), 500


@app.route('/chain', methods=['GET'])
def get_chain():
    chain_snapshot = blockchain.chain
    dict_chain = [block.__dict__.copy() for block in chain_snapshot]
    for dict_block in dict_chain:
        dict_block['transactions'] = [
            tx.__dict__ for tx in dict_block['transactions']]
    return jsonify(dict_chain), 200

@app.route('/broadcast-transaction',methods=['POST'])
@require_json
def broadcast_tranaction():
    values=request.get_json()
    if not values:
        response={'message':'Not Data Found'}
        return jsonify(response),400
    required=['sender','recipient','amount','signature']
    if not all(key in values for key in required):
        response={'message':'Some Data is Missing'}
        return jsonify(response),422     
    success=blockchain.add_transaction(values['recipient'],values['sender'],values['signature'],values['amount'])  
    if success:
            response = {
            'message': 'Successfully added tranaction.', 'transaction': {
                'sender': values['sender'], 'recipient': values['recipient'], 'amount': values['amount'], 'signature': values['signature']
            },
            'funds': blockchain.get_balance()
            }
            return jsonify(response), 201
        
@app.route('/transaction', methods=['GET'])
def get_open_transaction():
    transactions = blockchain.get_open_transactions()
    dict_transaction = [tx.__dict__ for tx in transactions]
    return jsonify(dict_transaction), 200


@app.route('/transaction', methods=['POST'])
@require_json
def add_transaction():
    if wallet.public_key == None:
        response = {
            'message': 'Adding a block failed.',
        }
        return jsonify(response), 401
    values = request.get_json()
    if not values:
        response = {
            'message': 'No Data found.'
        }
        return jsonify(response), 402
    require_fields = ['recipient', 'amount']
    if not all(field in values for field in require_fields):
        response = {
            'message': 'Require Data is missing.'
        }
        return jsonify(response), 402
    recipient = values['recipient']
    amount = values['amount']
    signature = wallet.sign_transaction(wallet.public_key, recipient, amount)
    success = blockchain.add_transaction(
        recipient, wallet.public_key, signature, amount)
    print(success, "this is where have bug")
    if success:
        response = {
            'message': 'Successfully added tranaction.', 'transaction': {
                'sender': wallet.public_key, 'recipient': recipient, 'amount': amount, 'signature': signature
            },
            'funds': blockchain.get_balance()
        }
        return jsonify(response), 201
    else:
        response = {
            'message': 'Creating a transaction failed'
        }
        return jsonify(response), 500


@app.route('/mine', methods=['POST'])
def mine():
    block = blockchain.mine_block()
    if block != None:
        dict_block = block.__dict__.copy()
        dict_block['transactions'] = [
            tx.__dict__ for tx in dict_block['transactions']]
        response = {
            'message': 'Block added successfully.',
            'block': dict_block,
            'funds': blockchain.get_balance()
        }
        return jsonify(response), 201
    else:
        response = {
            'message': 'Adding a block failed.',
            'wallet_set_up': wallet.public_key != None

        }
        return jsonify(response), 500


@app.route('/node', methods=['POST'])
@require_json
def add_node():
    values = request.get_json()
    if not values:
        response = {
            'message': 'No Data attached.'
        }
        return jsonify(response), 400

    if 'node' not in values:
        response = {
            'message': 'No NODE Data attached.'
        }
        return jsonify(response), 404
    node = values['node']
    blockchain.add_peer_node(node)
    response = {
        'message': 'Node Added successfully.',
        'all_nodes': blockchain.get_peer_node()
    }
    return jsonify(response), 201


@app.route('/node/<node_url>', methods=['DELETE'])
def remove_node(node_url):
    if node_url == '' or node_url == None:
        response = {
            'message': 'No Node found.'
        }
        return jsonify(response), 404
    blockchain.remove_peer_node(node_url)
    response = {
        'message': 'Node Removed.',
        'all_nodes': blockchain.get_peer_node()
    }
    return jsonify(response), 200


@app.route('/nodes', methods=['GET'])
def get_nodes():
    response = {
        'all_nodes': blockchain.get_peer_node()
    }
    return jsonify(response), 200


if __name__ == '__main__':
    from argparse import ArgumentParser
    parser=ArgumentParser()
    parser.add_argument('-p','--port',type=int,default=8000)
    args=parser.parse_args()
    print(args.port)
    port=args.port
    wallet = Wallet(port)
    blockchain = Blockchain(wallet.public_key,port)
    app.run(host='0.0.0.0', port=port)

File: static/js/node.js
new Vue({
  el: "#app",
  data: {
    blockchain: [],
    openTransactions: [],
    wallet: null,
    view: "chain",
    walletLoading: false,
    txLoading: false,
    dataLoading: false,
    showElement: null,
    error: null,
    success: null,
    funds: 0,
    outgoingTx: {
      recipient: "",
      amount: 0,
    },
  },
  computed: {
    loadedData: function () {
      return this.view === "chain" ? this.blockchain : this.openTransactions;
    },
  },
  methods: {
    onCreateWallet() {
      const vm = this;
      vm.walletLoading = true;
      axios
        .post("/wallet")
        .then((res) => {
          vm.error = null;
          vm.success = `Created Wallet! Public Key: ${res.data.public_key}, Private Key: ${res.data.private_key}`;
          vm.wallet = {
            public_key: res.data.public_key,
            private_key: res.data.private_key,
          };
          vm.funds = res.data.funds;
          vm.walletLoading = false;
        })
        .catch((err) => {
          vm.success = null;
          vm.error = err.response?.data?.message || "Error creating wallet.";
          vm.wallet = null;
          vm.walletLoading = false;
        });
    },
    onLoadWallet() {
      const vm = this;
      vm.walletLoading = true;
      axios
        .get("/wallet")
        .then((res) => {
          vm.error = null;
          vm.success = `Loaded Wallet! Public Key: ${res.data.public_key}, Private Key: ${res.data.private_key}`;
          vm.wallet = {
            public_key: res.data.public_key,
            private_key: res.data.private_key,
          };
          vm.funds = res.data.funds;
          vm.walletLoading = false;
        })
        .catch((err) => {
          vm.success = null;
          vm.error = err.response?.data?.message || "Error loading wallet.";
          vm.wallet = null;
          vm.walletLoading = false;
        });
    },
    onSendTx() {
      const vm = this;
      vm.txLoading = true;
      axios
        .post("/transaction", {
          recipient: vm.outgoingTx.recipient,
          amount: vm.outgoingTx.amount,
        })
        .then((res) => {
          vm.success = res.data.message;
          vm.error = null;
          vm.funds = res.data.funds;
          vm.txLoading = false;
        })
        .catch((err) => {
          vm.success = null;
          vm.error = err.response?.data?.message || "Transaction failed.";
          vm.txLoading = false;
        });
    },
    onMine() {
      const vm = this;
      axios
        .post("/mine")
        .then((res) => {
          vm.success = res.data.message;
          vm.error = null;
          vm.funds = res.data.funds;
        })
        .catch((err) => {
          vm.success = null;
          vm.error = err.response?.data?.message || "Mining failed.";
        });
    },
    onResolve() {
      const vm = this;
      axios
        .post("/resolve-conflicts")
        .then((res) => {
          vm.success = res.data.message;
          vm.error = null;
        })
        .catch((err) => {
          vm.success = null;
          vm.error = err.response?.data?.message || "Conflict resolution failed.";
        });
    },
    onLoadData() {
      const vm = this;
      vm.dataLoading = true;
      const url = this.view === "chain" ? "/chain" : "/transactions";
      axios
        .get(url)
        .then((res) => {
          if (vm.view === "chain") {
            vm.blockchain = res.data;
          } else {
            vm.openTransactions = res.data;
          }
          vm.dataLoading = false;
        })
        .catch(() => {
          vm.error = "Something went wrong.";
          vm.dataLoading = false;
        });
    },
  },
});
